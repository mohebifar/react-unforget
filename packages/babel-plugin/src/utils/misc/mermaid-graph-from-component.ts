import { encode } from "html-entities";
import type { Component } from "~/models/Component";
import type { ComponentSegment } from "~/models/segment/ComponentSegment";

export function mermaidGraphFromComponent(component: Component) {
  const rootSegment = component.getRootSegment();

  if (!rootSegment) {
    return "";
  }

  return printMermaidGraph(rootSegment);
}

function printMermaidGraph(root: ComponentSegment): string {
  const visited = new Map<ComponentSegment, string>();
  const graphLines: string[] = [
    "graph TD;",
    "classDef blockSegment fill:#5352ed,stroke:#333,stroke-width:2px;",
    "classDef rootChild stroke:#ff0000,stroke-width:2px;",
  ];
  let nodeIdCounter = 0;
  let subgraphCounter = 0;

  function getOrCreateNodeId(segment: ComponentSegment): string {
    if (!visited.has(segment)) {
      visited.set(segment, `node${++nodeIdCounter}`);
      const label = encode(segment.printCode().replace(/"/g, '\\"'));
      graphLines.push(
        `${visited.get(segment)}["<pre align="left">${label}</pre>"]`
      );
    }
    return visited.get(segment)!;
  }

  function traverse(node: ComponentSegment) {
    const nodeId = getOrCreateNodeId(node);
    const isRootChild = node.getParent() === root;

    if (node.getChildren().size > 0) {
      const subgraphId = `SG${++subgraphCounter}`;
      graphLines.push(`subgraph ${subgraphId}`);
      for (const child of node.getChildren()) {
        traverse(child);
      }
      graphLines.push("end");
      graphLines.push(`${nodeId} -.-> ${subgraphId}`);
      graphLines.push(`class ${nodeId} blockSegment`);
    }

    if (isRootChild) {
      graphLines.push(`class ${nodeId} rootChild`);
    }

    for (const dep of node.getDependencies()) {
      const depId = getOrCreateNodeId(dep.segment);
      graphLines.push(`${depId} --> ${nodeId}`);
    }
  }

  traverse(root);

  return graphLines.join("\n");
}
