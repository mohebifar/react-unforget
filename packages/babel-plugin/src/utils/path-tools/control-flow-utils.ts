type ControlFlowStatement =
  | babel.NodePath<babel.types.IfStatement>
  | babel.NodePath<babel.types.WhileStatement>
  | babel.NodePath<babel.types.DoWhileStatement>
  | babel.NodePath<babel.types.ForStatement>
  | babel.NodePath<babel.types.ForXStatement>
  | babel.NodePath<babel.types.TryStatement>;

export function getArgumentOfControlFlowStatement(
  statement: babel.NodePath<babel.types.Statement>,
) {
  if (statement.isIfStatement()) {
    return [statement.get("test")];
  }

  if (statement.isWhileStatement()) {
    return [statement.get("test")];
  }

  if (statement.isDoWhileStatement()) {
    return [statement.get("test")];
  }

  if (statement.isForStatement()) {
    return [
      statement.get("init"),
      statement.get("test"),
      statement.get("update"),
    ].filter(
      (s) => s.isExpression() || s.isVariableDeclaration(),
    ) as babel.NodePath<
      babel.types.Expression | babel.types.VariableDeclaration
    >[];
  }

  if (statement.isForOfStatement()) {
    return [statement.get("right")];
  }

  if (statement.isForInStatement()) {
    return [statement.get("right")];
  }

  return null;
}

export function isControlFlowStatement(
  statement: babel.NodePath,
): statement is ControlFlowStatement {
  return (
    statement.isIfStatement() ||
    statement.isWhileStatement() ||
    statement.isDoWhileStatement() ||
    statement.isForStatement() ||
    statement.isForXStatement() ||
    statement.isTryStatement()
  );
}

export function getControlFlowBodies(
  statement: babel.NodePath<babel.types.Statement>,
) {
  if (statement.isTryStatement()) {
    const block = statement.get("block");

    if (block.isBlockStatement()) {
      return [block];
    }
  }

  if (statement.isSwitchStatement()) {
    const cases = statement.get("cases");

    const blockStatements: babel.NodePath<babel.types.BlockStatement>[] = [];

    cases.forEach((c) => {
      const consequent = c.get("consequent");

      if (consequent.length > 0) {
        consequent.forEach((consequentStatement) => {
          if (consequentStatement.isBlockStatement()) {
            blockStatements.push(consequentStatement);
          }
        });
      }
    });

    return blockStatements;
  }

  if (statement.isIfStatement()) {
    const alternate = statement.get("alternate");
    const consequent = statement.get("consequent");

    return [consequent].concat(alternate.isStatement() ? [alternate] : []);
  }

  // I had to repeat the if statement here because of an issue in the type guarding
  if (statement.isWhileStatement()) {
    return [statement.get("body")];
  }

  if (statement.isDoWhileStatement()) {
    return [statement.get("body")];
  }

  if (statement.isForStatement()) {
    return [statement.get("body")];
  }

  if (statement.isForXStatement()) {
    return [statement.get("body")];
  }

  return [];
}

export function getBlockStatementsOfPath(
  path: babel.NodePath<babel.types.Statement>,
) {
  const bodies = getControlFlowBodies(path);

  return bodies.filter(
    (body): body is babel.NodePath<babel.types.BlockStatement> => {
      return body.isBlockStatement();
    },
  );
}

export function isForStatementInit(
  potentialInit: babel.NodePath<babel.types.Node>,
) {
  const forParent = potentialInit.findParent((p) =>
    p.isForStatement(),
  ) as babel.NodePath<babel.types.ForStatement> | null;

  if (!forParent) {
    return false;
  }

  const variableDeclaration = potentialInit.findParent((p) =>
    p.isVariableDeclaration(),
  ) as babel.NodePath<babel.types.VariableDeclaration> | null;

  if (!variableDeclaration) {
    return false;
  }

  return variableDeclaration === forParent.get("init");
}
