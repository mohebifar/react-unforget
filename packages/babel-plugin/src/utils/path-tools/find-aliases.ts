import type { Binding } from "@babel/traverse";
import * as t from "@babel/types";

export function findAliases(
  nodePath: babel.NodePath,
  filterFn: (path: babel.NodePath) => boolean,
) {
  const map = new Map<string, Binding[]>();

  nodePath.traverse({
    AssignmentExpression(assignmentPath) {
      const left = assignmentPath.get("left");
      const right = assignmentPath.get("right");
      if (!left.isIdentifier()) return;
      const name = left.node.name;
      map.set(name, extractAliasesFromRight(right, filterFn));
    },
    VariableDeclarator(declaratorPath) {
      const id: babel.NodePath = declaratorPath.get("id");
      const init = declaratorPath.get("init");

      if (!init.isExpression()) return;

      const name = t.isIdentifier(id.node) ? id.node.name : "unknown";

      map.set(name, extractAliasesFromRight(init, filterFn));
    },
  });

  return map;
}

function extractAliasesFromRight(
  nodePath: babel.NodePath<t.Expression>,
  filterFn: (path: babel.NodePath) => boolean,
) {
  const aliases: Binding[] = [];

  const handleFunction = (fn: babel.NodePath<t.Function>) => {
    const returnExpressions: babel.NodePath<t.Expression>[] = [];
    const body = fn.get("body");
    if (body.isBlockStatement()) {
      fn.traverse({
        ReturnStatement(returnPath) {
          const argumentPath = returnPath.get("argument");
          if (argumentPath.isExpression()) {
            returnExpressions.push(argumentPath);
          }
        },
      });
    } else if (body.isExpression()) {
      returnExpressions.push(body);
    }

    returnExpressions.forEach((returnExpression) => {
      handleExpression(returnExpression);
    });
  };

  const handleExpression = (expressionPath: babel.NodePath<t.Expression>) => {
    if (!expressionPath) return;

    switch (true) {
      case expressionPath.isIdentifier():
        // Direct alias
        const binding = nodePath.scope.getBinding(expressionPath.node.name);

        if (!binding || !filterFn(binding.path)) return;

        aliases.push(binding);
        // Check if this identifier is bound to a function, then evaluate that function
        const bindingPath = binding.path;
        if (bindingPath.isVariableDeclarator()) {
          const initPath = bindingPath.get("init");
          if (initPath.isExpression()) {
            handleExpression(initPath);
          }
        }
        break;
      case expressionPath.isCallExpression():
        // Evaluate the call expression to see if it returns a direct reference
        const calleePath = expressionPath.get("callee");
        if (calleePath.isIdentifier()) {
          const calleeBinding = nodePath.scope.getBinding(calleePath.node.name);
          if (calleeBinding && calleeBinding.path) {
            const calleePath = calleeBinding.path;

            let fn: babel.NodePath<t.Function> | null = null;

            if (calleePath.isFunction()) {
              fn = calleePath;
            }

            if (calleePath.isVariableDeclarator()) {
              const init = calleePath.get("init");
              if (init.isFunction()) {
                fn = init;
              }
            }

            if (fn) {
              handleFunction(fn);
            }
          }
        } else if (calleePath.isFunction()) {
          handleFunction(calleePath);
        }
        break;
      case expressionPath.isConditionalExpression():
        handleExpression(expressionPath.get("consequent"));
        handleExpression(expressionPath.get("alternate"));
        break;
      case expressionPath.isLogicalExpression():
        {
          if (expressionPath.node.operator === "&&") {
            handleExpression(expressionPath.get("right"));
          } else {
            handleExpression(expressionPath.get("left"));
            handleExpression(expressionPath.get("right"));
          }
        }
        break;
    }
  };

  handleExpression(nodePath);

  return aliases;
}
