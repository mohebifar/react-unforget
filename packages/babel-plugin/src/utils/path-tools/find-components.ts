import type * as babel from "@babel/core";
import * as t from "@babel/types";
import { Component } from "~/models/Component";
import { doesMatchHookName } from "~/utils/ast-tools/is-hook-call";
import type { TransformationContext } from "~/models/TransformationContext";
import { getReturnsOfFunction } from "./get-returns-of-function";

export function findComponents(
  program: babel.NodePath<babel.types.Program>,
  transformationContext?: TransformationContext,
) {
  const components: Component[] = [];

  program.traverse({
    Function(path) {
      let id: babel.types.Identifier | null | undefined = null;

      if (path.isFunctionDeclaration()) {
        id = path.node.id;
      }

      if (
        (path.isFunctionExpression() || path.isArrowFunctionExpression()) &&
        path.parentPath.isVariableDeclarator()
      ) {
        const varIdPath = path.parentPath.get("id");
        if (varIdPath.isIdentifier()) {
          id = varIdPath.node;
        }
      }

      const name = id?.name;
      const nameMatchesComponentName = name && doesIdMatchComponentName(name);
      const nameMatchesHook = name && doesMatchHookName(name);

      if (!nameMatchesComponentName && !nameMatchesHook && !name) {
        return;
      }

      if (transformationContext?.shouldSkipComponent(name)) {
        return;
      }

      if (path.isArrowFunctionExpression()) {
        const body = path.get("body");
        if (!body.isBlockStatement() && isComponentReturnType(body.node)) {
          components.push(
            new Component(
              path as babel.NodePath<babel.types.Function>,
              name,
              transformationContext,
            ),
          );
          return;
        }
      }

      const returns = getReturnsOfFunction(path);

      if (returns.length === 0) {
        return;
      }

      if (nameMatchesHook) {
        components.push(
          new Component(
            path as babel.NodePath<babel.types.Function>,
            name,
            transformationContext,
          ),
        );
      }

      const allReturnsMatch = returns.every((returnStatement) => {
        const argument = returnStatement.get("argument");
        if (isComponentReturnType(argument.node)) {
          return true;
        }

        // TODO: Recursively follow re-assignments of the variable
        if (argument.isIdentifier()) {
          const binding = argument.scope.getBinding(argument.node.name);
          const variableDeclarator = binding?.path.find((bindingPath) =>
            bindingPath.isVariableDeclarator(),
          ) as babel.NodePath<babel.types.VariableDeclarator> | undefined;

          if (variableDeclarator) {
            const init = variableDeclarator.node;
            if (init && isComponentReturnType(init)) {
              return true;
            }
          }
        }
      });

      if (allReturnsMatch) {
        components.push(
          new Component(
            path as babel.NodePath<babel.types.Function>,
            name,
            transformationContext,
          ),
        );
      }
    },
  });

  return components;
}

function doesIdMatchComponentName(name: string) {
  return /^_?[A-Z]/.test(name);
}

function isComponentReturnType(node: babel.types.Node | null | undefined) {
  return t.isJSXElement(node) || t.isJSXFragment(node) || t.isNullLiteral(node);
}
