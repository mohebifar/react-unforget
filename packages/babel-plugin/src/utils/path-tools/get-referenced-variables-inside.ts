import type * as babel from "@babel/core";
import type { Binding } from "@babel/traverse";

export function getReferencedVariablesInside(
  path: babel.NodePath<babel.types.Node>,
  unique = true,
) {
  const visited = new Set<Binding>();
  const map = new Map<
    babel.NodePath<babel.types.Identifier | babel.types.JSXIdentifier>,
    Binding
  >();

  path.traverse({
    ReferencedIdentifier(innerPath) {
      const name = innerPath.node.name;
      const binding = path.scope.getBinding(name);
      if (binding && (!unique || !visited.has(binding))) {
        visited.add(binding);
        map.set(innerPath, binding);
      }
    },
  });

  return map;
}
