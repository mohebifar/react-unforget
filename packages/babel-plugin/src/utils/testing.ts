import * as fs from "fs";
import * as path from "path";
import * as babel from "@babel/core";
import * as generateBase from "@babel/generator";
import traverse from "@babel/traverse";
import type * as t from "@babel/types";
import { visitProgram } from "~/visit-program";
import { mermaidGraphFromComponent } from "~/utils/misc/mermaid-graph-from-component";
import babelPlugin from "../index";

const babelTransformOptions = {
  plugins: [
    ["@babel/plugin-transform-typescript", { isTSX: true }],
    babelPlugin,
  ],
} satisfies babel.TransformOptions;

export function transformWithCore(input: string) {
  const { code } = babel.transform(input, babelTransformOptions)!;
  return code;
}

export function transformForJest(input: string, filename: string) {
  const { code } = babel.transformSync(input, {
    ...babelTransformOptions,
    sourceMaps: "both",
    compact: false,
    presets: [["@babel/preset-react", { runtime: "automatic" }], "jest"],
    filename,
    plugins: [
      ...babelTransformOptions.plugins,
      "@babel/plugin-transform-modules-commonjs",
    ],
  })!;
  return code;
}

export function transformWithParseAndCast(input: string) {
  const root = parse(input);

  visitProgram(root);

  return String(root);
}

export function parse(input: string) {
  const ast = babel.parse(input, babelTransformOptions)!;

  let result: babel.NodePath<t.Program> | null = null;
  traverse(ast, {
    enter(path) {
      result = path as babel.NodePath<t.Program>;
      path.skip();
    },
  });

  return result!;
}

export function loadFixture(fixturePath: string) {
  const fixture = path.resolve(
    __dirname,
    "../tests/fixtures",
    fixturePath + ".tsx"
  );

  const fileContent = fs.readFileSync(fixture, "utf-8");

  return fileContent;
}

export function parseFixture(fixturePath: string) {
  return parse(loadFixture(fixturePath));
}

export function generate(path: babel.types.Node) {
  return generateBase.default(path).code;
}

export { mermaidGraphFromComponent };
