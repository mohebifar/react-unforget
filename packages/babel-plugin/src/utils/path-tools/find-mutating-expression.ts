import { MUTATING_METHODS } from "~/utils/constants";
import { getLeftmostIdName } from "~/utils/ast-tools/get-leftmost-id-name";
import { getRightmostIdName } from "~/utils/ast-tools/get-rightmost-id-name";
import type { Binding } from "@babel/traverse";
import { getDeclaredIdentifiersInLVal } from "./get-declared-identifiers-in-lval";
import { getReferencedVariablesInside } from "./get-referenced-variables-inside";

export function findMutatingExpressions(
  path: babel.NodePath<babel.types.Node>
) {
  const result: {
    binding: Binding;
    path: babel.NodePath<babel.types.Expression>;
  }[] = [];

  const referencedIds = getReferencedVariablesInside(path).entries();

  for (const [referencePath, binding] of referencedIds) {
    const name = binding.identifier.name;

    const found = referencePath.find((currentParentPath) => {
      if (currentParentPath.isStatement()) {
        return true;
      }

      if (currentParentPath.isAssignmentExpression()) {
        const left = currentParentPath.node.left;
        const leftMostIds = currentParentPath.isLVal()
          ? getDeclaredIdentifiersInLVal(currentParentPath)
          : [getLeftmostIdName(left)];

        return leftMostIds.includes(name);
      }

      if (currentParentPath.isUpdateExpression()) {
        return getLeftmostIdName(currentParentPath.node.argument) === name;
      }

      if (currentParentPath.isCallExpression()) {
        const callee = currentParentPath.get("callee");

        if (callee.isMemberExpression()) {
          const leftMostId = getLeftmostIdName(callee.node);
          const rightmostId = getRightmostIdName(callee.node);

          if (MUTATING_METHODS.includes(rightmostId)) {
            return leftMostId === name;
          }
        }
      }

      return false;
    });

    if (found?.isExpression()) {
      result.push({ binding, path: found });
    }
  }

  return result;
}
