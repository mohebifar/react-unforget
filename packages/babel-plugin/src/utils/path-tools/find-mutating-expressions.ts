import type { Binding } from "@babel/traverse";
import { getLeftmostIdName } from "~/utils/ast-tools/get-leftmost-id-name";
import { getRightmostIdName } from "~/utils/ast-tools/get-rightmost-id-name";
import { MUTATING_METHODS } from "~/utils/constants";
import { getDeclaredIdentifiersInLVal } from "./get-declared-identifiers-in-lval";

export function findMutatingExpressions(
  path: babel.NodePath<babel.types.Node>,
) {
  const result: {
    binding: Binding;
    path: babel.NodePath<babel.types.Expression>;
  }[] = [];

  const pairExists = (
    binding: Binding,
    path: babel.NodePath<babel.types.Node>,
  ) => result.some((entry) => entry.binding === binding && entry.path === path);

  const push = (
    binding: Binding,
    path: babel.NodePath<babel.types.Expression>,
  ) => {
    if (!pairExists(binding, path)) {
      result.push({ binding, path });
    }
  };

  path.traverse({
    AssignmentExpression(assignmentPath) {
      const lval = assignmentPath.get("left");
      if (!lval.isLVal()) {
        return;
      }

      const statement = assignmentPath.getStatementParent();
      const leftMostIds = lval.isMemberExpression()
        ? [getLeftmostIdName(lval.node)]
        : getDeclaredIdentifiersInLVal(lval);

      if (leftMostIds.length === 0 || !statement) {
        return;
      }

      leftMostIds.forEach((name) => {
        const binding = statement.scope.getBinding(name);

        if (binding) {
          push(binding, assignmentPath);
        }
      });
    },
    UpdateExpression(updatePath) {
      const argument = updatePath.get("argument");

      const leftMostId = getLeftmostIdName(argument.node);

      const binding = argument.scope.getBinding(leftMostId);

      if (binding) {
        push(binding, updatePath);
      }
    },
    CallExpression(callPath) {
      const callee = callPath.get("callee");

      if (!callee.isMemberExpression()) {
        return;
      }
      try {
        const leftMostId = getLeftmostIdName(callee.node);
        const rightmostId = getRightmostIdName(callee.node);

        if (MUTATING_METHODS.includes(rightmostId)) {
          const binding = callee.scope.getBinding(leftMostId);

          if (binding) {
            push(binding, callPath);
          }
        }
      } catch {
        // The callee was not a member expression or identifier
      }
    },
  });

  return result;
}
