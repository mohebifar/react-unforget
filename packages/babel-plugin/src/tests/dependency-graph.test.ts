import { mermaidGraphFromComponent } from "~/utils/misc/mermaid-graph-from-component";
import { findComponents } from "~/utils/path-tools/find-components";
import { parseFixture } from "~/utils/testing";

describe("dependency graph", () => {
  it("fixture_dependency_graph", () => {
    const path = parseFixture("fixture_dependency_graph");
    const components = findComponents(path);

    const mermaidGraphs: string[] = [];

    const componentCodes: string[] = [];

    components.forEach((component) => {
      component.analyze();

      mermaidGraphs.push(mermaidGraphFromComponent(component));
      componentCodes.push(component.path.toString());
    });

    expect(mermaidGraphs).toMatchSnapshot();
    expect(componentCodes).toMatchSnapshot();
  });
});
