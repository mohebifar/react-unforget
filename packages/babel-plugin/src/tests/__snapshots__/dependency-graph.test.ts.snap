// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependency graph fixture_dependency_graph 1`] = `
[
  "graph TD;
classDef blockSegment fill:#5352ed,stroke:#333,stroke-width:2px;
classDef rootChild stroke:#ff0000,stroke-width:2px;
classDef returnNetwork stroke:#2ecc71,stroke-width:2px;
node1["<pre align="left">function Test(_props) {
  let a = _props.a;
  let b = _props.b;
  const _unwrapped = useState(0);
  const state = _unwrapped[0];
  const setState = _unwrapped[1];
  let x: number[] = [];
  const z: number[] = [];
  let y = state % 2 === 0 ? x : z;
  y.push(state);
  let _unwrapped2 = 10;
  const _unwrapped3 = [5, 6];
  let _unwrapped4 = _unwrapped3[0];
  let _unwrapped5 = _unwrapped3[1];
  for (let i = _unwrapped2, j = _unwrapped4, k = _unwrapped5; j &lt; i; j++) {
    y.push(i + j * b);
  }
  let _unwrapped6 = 10;
  for (let i = _unwrapped6; i &lt; 10; i++) {
    y.push(i);
  }
  const m = y;
  let _unwrapped7 = state;
  for (let i = _unwrapped7; i &lt; 10; i++) {
    m.push(i);
  }
  const _unwrappedJsxExp = y[0];
  const _unwrappedJsxExp2 = y[0];
  const _unwrappedJsxEl = &lt;button&gt;Click here&lt;/button&gt;;
  const _unwrappedJsxEl2 = &lt;div&gt;
      {state} {_unwrappedJsxExp} {a} {_unwrappedJsxEl} {_unwrappedJsxExp2}
    &lt;/div&gt;;
  return _unwrappedJsxEl2;
}</pre>"]
subgraph SG1
node2["<pre align="left">_props</pre>"]
class node2 returnNetwork;
class node2 rootChild
node3["<pre align="left">{
  let a = _props.a;
  let b = _props.b;
  const _unwrapped = useState(0);
  const state = _unwrapped[0];
  const setState = _unwrapped[1];
  let x: number[] = [];
  const z: number[] = [];
  let y = state % 2 === 0 ? x : z;
  y.push(state);
  let _unwrapped2 = 10;
  const _unwrapped3 = [5, 6];
  let _unwrapped4 = _unwrapped3[0];
  let _unwrapped5 = _unwrapped3[1];
  for (let i = _unwrapped2, j = _unwrapped4, k = _unwrapped5; j &lt; i; j++) {
    y.push(i + j * b);
  }
  let _unwrapped6 = 10;
  for (let i = _unwrapped6; i &lt; 10; i++) {
    y.push(i);
  }
  const m = y;
  let _unwrapped7 = state;
  for (let i = _unwrapped7; i &lt; 10; i++) {
    m.push(i);
  }
  const _unwrappedJsxExp = y[0];
  const _unwrappedJsxExp2 = y[0];
  const _unwrappedJsxEl = &lt;button&gt;Click here&lt;/button&gt;;
  const _unwrappedJsxEl2 = &lt;div&gt;
      {state} {_unwrappedJsxExp} {a} {_unwrappedJsxEl} {_unwrappedJsxExp2}
    &lt;/div&gt;;
  return _unwrappedJsxEl2;
}</pre>"]
subgraph SG2
node4["<pre align="left">let x: number[] = [];</pre>"]
class node4 returnNetwork;
node5["<pre align="left">const z: number[] = [];</pre>"]
class node5 returnNetwork;
node6["<pre align="left">let y = state % 2 === 0 ? x : z;</pre>"]
class node6 returnNetwork;
node7["<pre align="left">const state = _unwrapped[0];</pre>"]
class node7 returnNetwork;
node7 --> node6
node4 --> node6
node5 --> node6
node8["<pre align="left">y.push(state);</pre>"]
node8 -.-> node6
node9["<pre align="left">y.push(i + j * b);</pre>"]
node9 -.-> node6
node10["<pre align="left">y.push(i);</pre>"]
node10 -.-> node6
node11["<pre align="left">const _unwrapped = useState(0);</pre>"]
class node11 returnNetwork;
node11 --> node7
node12["<pre align="left">const setState = _unwrapped[1];</pre>"]
node6 --> node8
node7 --> node8
node13["<pre align="left">for (let i = _unwrapped2, j = _unwrapped4, k = _unwrapped5; j &lt; i; j++) {
  y.push(i + j * b);
}</pre>"]
subgraph SG3
node14["<pre align="left">let i = _unwrapped2,
  j = _unwrapped4,
  k = _unwrapped5;</pre>"]
node15["<pre align="left">let _unwrapped2 = 10;</pre>"]
node15 --> node14
node16["<pre align="left">let _unwrapped4 = _unwrapped3[0];</pre>"]
node16 --> node14
node17["<pre align="left">let _unwrapped5 = _unwrapped3[1];</pre>"]
node17 --> node14
node18["<pre align="left">{
  y.push(i + j * b);
}</pre>"]
subgraph SG4
node6 --> node9
node14 --> node9
node14 --> node9
node19["<pre align="left">let b = _props.b;</pre>"]
node19 --> node9
end
node18 -.-> SG4
class node18 blockSegment
end
node13 -.-> SG3
class node13 blockSegment
node14 --> node13
node14 --> node13
node20["<pre align="left">const _unwrapped3 = [5, 6];</pre>"]
node20 --> node16
node20 --> node17
node21["<pre align="left">let a = _props.a;</pre>"]
class node21 returnNetwork;
node2 --> node21
node2 --> node19
node22["<pre align="left">for (let i = _unwrapped6; i &lt; 10; i++) {
  y.push(i);
}</pre>"]
subgraph SG5
node23["<pre align="left">let i = _unwrapped6;</pre>"]
node24["<pre align="left">let _unwrapped6 = 10;</pre>"]
node24 --> node23
node25["<pre align="left">{
  y.push(i);
}</pre>"]
subgraph SG6
node6 --> node10
node23 --> node10
end
node25 -.-> SG6
class node25 blockSegment
end
node22 -.-> SG5
class node22 blockSegment
node23 --> node22
node26["<pre align="left">const m = y;</pre>"]
node6 --> node26
node27["<pre align="left">m.push(i);</pre>"]
node27 -.-> node26
node28["<pre align="left">for (let i = _unwrapped7; i &lt; 10; i++) {
  m.push(i);
}</pre>"]
subgraph SG7
node29["<pre align="left">let i = _unwrapped7;</pre>"]
node30["<pre align="left">let _unwrapped7 = state;</pre>"]
node30 --> node29
node31["<pre align="left">{
  m.push(i);
}</pre>"]
subgraph SG8
node26 --> node27
node29 --> node27
end
node31 -.-> SG8
class node31 blockSegment
end
node28 -.-> SG7
class node28 blockSegment
node7 --> node28
node29 --> node28
node7 --> node30
node32["<pre align="left">const _unwrappedJsxExp = y[0];</pre>"]
class node32 returnNetwork;
node6 --> node32
node33["<pre align="left">const _unwrappedJsxExp2 = y[0];</pre>"]
class node33 returnNetwork;
node6 --> node33
node34["<pre align="left">const _unwrappedJsxEl = &lt;button&gt;Click here&lt;/button&gt;;</pre>"]
class node34 returnNetwork;
node35["<pre align="left">const _unwrappedJsxEl2 = &lt;div&gt;
      {state} {_unwrappedJsxExp} {a} {_unwrappedJsxEl} {_unwrappedJsxExp2}
    &lt;/div&gt;;</pre>"]
class node35 returnNetwork;
node7 --> node35
node32 --> node35
node21 --> node35
node34 --> node35
node33 --> node35
node36["<pre align="left">return _unwrappedJsxEl2;</pre>"]
class node36 returnNetwork;
node35 --> node36
end
node3 -.-> SG2
class node3 blockSegment
class node3 rootChild
end
node1 -.-> SG1
class node1 blockSegment",
]
`;

exports[`dependency graph fixture_dependency_graph 2`] = `
[
  "function Test(_props) {
  let a = _props.a;
  let b = _props.b;
  const _unwrapped = useState(0);
  const state = _unwrapped[0];
  const setState = _unwrapped[1];
  let x: number[] = [];
  const z: number[] = [];
  let y = state % 2 === 0 ? x : z;
  y.push(state);
  let _unwrapped2 = 10;
  const _unwrapped3 = [5, 6];
  let _unwrapped4 = _unwrapped3[0];
  let _unwrapped5 = _unwrapped3[1];
  for (let i = _unwrapped2, j = _unwrapped4, k = _unwrapped5; j < i; j++) {
    y.push(i + j * b);
  }
  let _unwrapped6 = 10;
  for (let i = _unwrapped6; i < 10; i++) {
    y.push(i);
  }
  const m = y;
  let _unwrapped7 = state;
  for (let i = _unwrapped7; i < 10; i++) {
    m.push(i);
  }
  const _unwrappedJsxExp = y[0];
  const _unwrappedJsxExp2 = y[0];
  const _unwrappedJsxEl = <button>Click here</button>;
  const _unwrappedJsxEl2 = <div>
      {state} {_unwrappedJsxExp} {a} {_unwrappedJsxEl} {_unwrappedJsxExp2}
    </div>;
  return _unwrappedJsxEl2;
}",
]
`;
